Mathlib/Logic/UnivLE.lean

```diff
@@ -24,31 +21,44 @@ noncomputable section
 /--
 A class expressing a universe inequality. `UnivLE.{u, v}` expresses that `u ≤ v`.
 
-There are (at least) two plausible definitions for `u ≤ v`:
-* strong: `∀ α : Type max u v, Small.{v} α`
-* weak: `∀ α : Type u, Small.{v} α`
-
-The weak definition has the advantage of being transitive.
-However only under the strong definition do we have `Small.{v} ((α : Type u) → (β : Type v))`,
-which is essential for proving that `Type v` has `Type u`-indexed limits when `u ≤ v`.
+There used a stronger definition `∀ α : Type max u v, Small.{v} α` that immediately implies
+`Small.{v} ((α : Type u) → (β : Type v))` which is essential for proving that `Type v` has
+`Type u`-indexed limits when `u ≤ v`. However the current weaker condition
+`∀ α : Type u, Small.{v} α` also implies the same, so we switched to use it for
+its simplicity and transitivity.
 
-The strong definition implies the weaker definition (see below),
+The strong definition easily implies the weaker definition (see below),
 but we can not prove the reverse implication.
 This is because in Lean's type theory, while `max u v` is at least at big as `u` and `v`,
 it could be bigger than both!
+See also `Mathlib.CategoryTheory.UnivLE` for the statement that the stronger definition is
+equivalent to `EssSurj (uliftFunctor : Type v ⥤ Type max u v)`.
 -/
 @[pp_with_univ]
-abbrev UnivLE.{u, v} : Prop := ∀ α : Type max u v, Small.{v} α
+abbrev UnivLE.{u, v} : Prop := ∀ α : Type u, Small.{v} α
 
 example : UnivLE.{u, u} := inferInstance
 example : UnivLE.{u, u+1} := inferInstance
 example : UnivLE.{0, u} := inferInstance
-example : UnivLE.{u, max u v} := inferInstance
+/- Why can't Lean infer the following instances? -/
+instance (priority := 100) : UnivLE.{u, max u v} := fun α ↦ small_max.{v} α
+instance (priority := 100) : UnivLE.{u, max v u} := fun α ↦ small_max.{v} α
```

Comment:
Probably that is causing typeclass inference failure here.

Reply:
Before, the asymmetry between `max u v` and `v` might have helped ...

Reply:
Unfortunately, it turns out I had an illusion and I can't prove (with the weaker definition) that `UnivLE.{u, v}` implies `UnivLE.{max u v, v}` after all, and I can't even show transitivity of the stronger definition. [All I can show](https://github.com/leanprover-community/mathlib4/pull/7695/commits/d3796a38310362ee1f497cd08123eba0fa47d8af#diff-831991b0359b52c44fff62c60b0a456e7e2b3453d8a9e23d9ee9b3813bd473ffR69-R81) is that `UnivLE.{u, v} ∨ UnivLE.{v, u}` so `UnivLE` "is" a total preorder on the universe levels, but even when `UnivLE.{u, v}` and `UnivLE.{v, u}` both hold you can't say that u and v are the same and substitute to get max u w "=" max v w, or even prove `UnivLE.{max u w, max v w}`. And it can't be ruled out that `Cardinal.univ.{max u v}` is bigger than `max univ.{u} univ.{v}`.

Lean's universe equality checking rules (see [leantt](https://github.com/digama0/lean-type-theory/releases/download/v1.0/main.pdf)) do appear to intend that `max u v` be the least upper bound of `u` and `v`: 
<img width="212" alt="image" src="https://github.com/leanprover-community/mathlib4/assets/3064145/f3a40c23-b8e5-4ef8-a969-4902c4817363">
and there's even an "induction principle":
<img width="408" alt="image" src="https://github.com/leanprover-community/mathlib4/assets/3064145/82acc9d2-7ea6-43bd-9071-824fec4a5c51">
but there's no way to take advantage of them. I wonder whether there's a characterization of equalities between universe expressions that are derived from these rules, and what non-standard models they admit.

Since we can't show the two definitions are equivalent and feel free to use the stronger definition, we must consider using the weaker definition which makes resolving [lean4#2297](https://github.com/leanprover/lean4/issues/2297) important. I'll try to fix things anyway and see how far I can get without a functioning `UnivLE.{u, max u v}` instance (it has been [pointed out](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.233463.20universe.20constraint.20issues/near/354705892
) by @adamtopaz that this will be problematic.)